### PR TITLE
NHSNLINK-381: Create DocumentReference for new reports

### PR DESCRIPTION
**Pull request by arianamihailescu**

### CODE REVIEW

File: api/src/main/java/com/lantanagroup/link/api/controller/BaseController.java [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
+import org.hl7.fhir.r4.model.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.security.core.Authentication;
 
 import javax.servlet.http.HttpServletRequest;
 
 public class BaseController {
+  private static final Logger logger = LoggerFactory.getLogger(PatientIdentifierController.class);
```
Comment (seanmcilvenna) [inline api/src/main/java/com/lantanagroup/link/api/controller/BaseController.java:16]: `PatientIdentifierController.class` should be `BaseController.class` here.

File: api/src/main/java/com/lantanagroup/link/api/controller/BaseController.java [hunk 2]
```diff
@@ -22,4 +27,28 @@ protected IGenericClient getFhirStoreClient(Authentication authentication, HttpS
...
+    if (!outcome.getCreated() || outcome.getResource() == null) {
+      logger.error("Failed to store/create FHIR resource");
+    } else {
+      logger.debug("Stored FHIR resource with new ID of " + outcome.getResource().getIdElement().getIdPart());
+    }
+  }
+
+  protected void updateResource(Resource resource, IGenericClient fhirStoreClient) {
+
+    if (resource.getMeta().getVersionId() == null){
```
Comment (seanmcilvenna) [inline api/src/main/java/com/lantanagroup/link/api/controller/BaseController.java:42]: Remove unused `if` block, or at least comment it out with a reason why

File: api/src/main/java/com/lantanagroup/link/api/controller/ReportController.java [hunk 3]
```diff
@@ -301,17 +301,71 @@ public QueryReport generateReport(@AuthenticationPrincipal LinkCredentials user,
...
       FhirHelper.recordAuditEvent(request, fhirStoreClient,  user.getJwt(), FhirHelper.AuditEventTypes.Generate, "Successfully Generated Report");
+
+      // Save measure report
+      String id = RandomStringUtils.randomAlphanumeric(8);
+      measureReport.setId(id);
+      updateResource(measureReport, fhirStoreClient);
+
+      // Save document reference
+      DocumentReference documentReference = getDocumentReference(user, contextData, id);
+      createResource(documentReference, fhirStoreClient);
```
Comment (seanmcilvenna) [inline api/src/main/java/com/lantanagroup/link/api/controller/ReportController.java:315]: I generally prefer `this.createResource()` syntax (explicitly using `this.` when calling class-instance methods). Not required, but please consider.

Comment (seanmcilvenna) [reply api/src/main/java/com/lantanagroup/link/api/controller/ReportController.java:315]: In this case, the same would apply for `this.createResource()` and `this.getDocumentReference()` and `this.updateMeasure()`

File: api/src/main/java/com/lantanagroup/link/api/controller/ReportController.java [hunk 4]
```diff
@@ -301,17 +301,71 @@ public QueryReport generateReport(@AuthenticationPrincipal LinkCredentials user,
...
     } catch (Exception ex) {
       logger.error(String.format("Error generating report: %s", ex.getMessage()), ex);
       throw new HttpResponseException(500, "Please contact system administrator regarding this error");
     }
-
     return report;
   }
 
+  private DocumentReference getDocumentReference (LinkCredentials user, HashMap<String, Object> contextData, String id) {
```
Comment (seanmcilvenna) [inline api/src/main/java/com/lantanagroup/link/api/controller/ReportController.java:323]: Consider renaming to `generateDocumentReference()`. I like to use
`getX()` for retrieving from a data store or for a getter
`createX()` to create a new model and persist it
`generateX()` to generate an instance of a model, but not persist it (yet)

**Review by seanmcilvenna — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/lantanagroup/link/api/controller/BaseController.java (hunk 1 above)
Comment (arianamihailescu) [reply api/src/main/java/com/lantanagroup/link/api/controller/BaseController.java]: Completed

**Review by seanmcilvenna — APPROVED** [review]

**Pull request merged**